### ISSUE TITLE
Add a lookup from state root to block root

### ISSUE THREAD

**Issue by ajsutton** [Consensys-Incorporated/teku#2184]
### Description
So that I can lookup states by state root, add a mapping to hot and finalised storage that stores a `state root -> (block root, slot)` mapping.  In finalised data store `state root -> slot`.

Ensure that mapping is added even for skipped slots (need to add an entry in `StateTransition.process_slot` when we set the new `state.state_roots` entry.

When given a state root, we can then look up the associated block root (either directly in hot states or via `state root -> slot -> block root` for finalised) and from that get the state.  The returned state will not have any empty slots processed, so we may need to reprocess those empty slots which is why we're storing the slot not just the block root.

**Comment by ajsutton** [comment]
We need to add a new lookup to `Store` which records `stateRoot->(blockRoot, slot)`. Using that information
we can regenerate any state by using the `blockRoot` to retrieve the state at the last block, then 
processing any empty slots to get up to `slot`.  We should ensure that any required processing is done
without holding the `Store.readLock`.

Can populate the state root lookup when regenerating states at startup (same time we populate `block_states`).
Will need to ensure we capture the root for empty slots here.

Q: We don't currently process slots until a block is received - should we change that to always track latest
head of the chain so we know the state root if the latest slot is empty?

`Store.Transaction` will need to record new state roots in `putBlockAndState`. It will also need
 a new `putStateRoot(blockRoot, slot)` to support recording the state root for empty slots.
 `StateTransition.process_slot` will need to be able to optionally call `putBlockAndState` for each
 slot it processes.  Since this is in the middle of a transition we need to test carefully to ensure
 we capture the state root, slot and block root correctly so they are all consistent.
 
`StoreTransactionUpdates` will need to be updated to send any new state roots to the storage system
and to prune state roots from prior to finalization. Iterating through all state roots to prune based
on the slot is probably reasonable.

The database won't need to store state roots for hot state as they can be regenerated. For finalized
states it can store `stateRoot -> slot` and then use the existing `slot -> blockRoot` lookup to get
the block root.

Detecting which state roots are on the finalized chain will have to be done by iterating through the 
pruned states, and recording any state which matches a finalized block root as finalized.

The way this integrates may change when the work to support storing only some blocks in memory lands so will revisit then to do the actual integration work.

**Comment by ajsutton** [comment]
So plan is:
1. Capture state root for each generated state similar to the `InterimStateListener` in https://github.com/PegaSysEng/teku/compare/master...ajsutton:state-root-to-block-root These will be added to the store transaction and stored to the hot database but not kept in memory.
2. Lookup by state root, goes to disk to load block root and slot for the given state root.  Then uses getStateByBlockRoot to get the closest state and reprocess any empty slots up to the required slot.
3. On finalisation, start a background thread to walk through each state root in the hot database and delete it if the slot has now been finalised.
4. For finalised data (if in prune mode, do nothing), start at the last finalised state we have stored, jump forward by `SLOTS_PER_HISTORICAL_ROOT` at a time and then use the state roots field from that state to store a stateRoot -> slot lookup.  We may need to generate states or may be able to just use states present in the store transaction.

Follow up: while walking through finalised states to store state root lookup, store any states we regenerated that weren't already available from the store transaction because they'd been pruned.
